# Prueba Chi Cuadrada

## Nombre: David Egas



Realice un programa que permita calcular el valor de Chi-Cuadrada y genere la gráfica de distribución de los 100 primeros números pseudo-aleatorios generados por los métodos de cuadrados medios y productos medios.

Emplee el siguiente nivel de significancia α=0.05


### Para primer método de numeros PseudoAleatorios

In [35]:
import collections
def divisionValores(long,numeroOriginal,digitos):
    numero=int((long/2)-(digitos/2)+1)
    sNumero=str(numeroOriginal)
    numUI=sNumero[numero-1:numero+3]
    return int(numUI)

In [36]:
def buscarSemilla(valor,lista):
    for x in range(0,len(lista)):
        if (lista[x]==valor):
               return True

In [38]:
listaFinal = []
Rn=0
def generarNumeroPseudoaleatorios(iteraciones,semilla):
    iteraciones=iteraciones
    Xo=semilla
    digitos=7
    Xn=0
    XnPo=0
    longuitud=0
    Ui=0
    
    arregloSemillas=[]
    semillasRepetidas=[]
    
    if len(str(Xo))>3:
        print('PROCESO NORMAL')
        for i in range(iteraciones):
            if (i==0):
                Xn=Xo
                XnPo=Xn**2
                longuitud=len(str(XnPo))
                Ui=divisionValores(longuitud,XnPo,digitos)
                arregloSemillas.append(Xn)
                Rn=Ui/10000
                print(Rn) 
               
            else:
                Xn=Ui
                XnPo=Xn**2
                longuitud=len(str(XnPo))
                Ui=divisionValores(longuitud,XnPo,digitos)
                Rn=Ui/10000
                print(Rn)
                listaFinal.append(Rn)
                
                busquedaSemilla=buscarSemilla(Xn,arregloSemillas)
                if(busquedaSemilla==True):
                    semillasRepetidas.append(Xn)
                    print('LA SEMILLA SE VUELE A REPETIR EN LA ITERACION:',i,'VALOR:',Xn)
                arregloSemillas.append(Xn)

    else:
        print('LA SEMILLA ES MENOR A 3')

    print('LA FRECUENCIA DE REPETICION ES DE:')
    #Vemos la frecuencia de repeticion
    counter=collections.Counter(semillasRepetidas)
    print(counter)

In [39]:
iteraciones=100
semilla=74731897457
generarNumeroPseudoaleatorios(iteraciones,semilla)

PROCESO NORMAL
0.4975
0.2475
0.6125
0.3751
0.1407
0.1979
0.3916
0.1533
0.235
0.5522
0.3049
0.9296
0.8641
0.7466
0.5574
0.3106
0.9647
0.9306
0.866
0.7499
0.5623
0.3161
0.9991
0.9982
0.9964
0.9928
0.9856
0.9714
0.9436
0.8903
0.7926
0.6282
0.3946
0.1557
0.2424
0.5875
0.3451
0.119
0.1416
0.2005
0.402
0.1616
0.2611
0.6817
0.4647
0.2159
0.4661
0.2172
0.4717
0.2225
0.495
0.245
0.6002
0.3602
0.1297
0.1682
0.2829
0.8003
0.6404
0.4101
0.1681
0.2825
0.798
0.6368
0.4055
0.1644
0.2702
0.73
0.5329
0.2839
0.8059
0.6494
0.4217
0.1778
0.3161
0.9991
LA SEMILLA SE VUELE A REPETIR EN LA ITERACION: 75 VALOR: 3161
0.9982
LA SEMILLA SE VUELE A REPETIR EN LA ITERACION: 76 VALOR: 9991
0.9964
LA SEMILLA SE VUELE A REPETIR EN LA ITERACION: 77 VALOR: 9982
0.9928
LA SEMILLA SE VUELE A REPETIR EN LA ITERACION: 78 VALOR: 9964
0.9856
LA SEMILLA SE VUELE A REPETIR EN LA ITERACION: 79 VALOR: 9928
0.9714
LA SEMILLA SE VUELE A REPETIR EN LA ITERACION: 80 VALOR: 9856
0.9436
LA SEMILLA SE VUELE A REPETIR EN LA ITERACION: 8

In [40]:
chisquare(lista, ddof=0.05)

Power_divergenceResult(statistic=1.162162162162162, pvalue=0.9968424171841995)

El valor estadístico es: 1.162162162162162 y el valor de prueba es 0.9968424171841995 por lo que los números siguen una distribución uniforme.

### Para el segundo  método de números PseudoAleatorios



In [42]:
import numpy as np
import time


rango = int((input("Ingrese la cantidad de numeros pseudoaleatorios:  ")))

a = 74731897457
b = 37747318974
m = 19

cont = 0
lista = []
contInterac = []

for n in range(rango):
    semilla = int(time.time())
    
    print("semilla",semilla,"iteracion",n)
    cont = 0
    while(True):
        
        xn = (a * semilla + b) % m
        semilla = xn 
        div = semilla / m
        
        #print(cont,'\t\t',semilla,'\t',"{0:.4f}".format(div),(semilla in lista) is True)
        
        if((div in lista) is True):
            
            contInterac.append(cont)
            
            break
       
        cont = cont+1
        lista.append(div)
        
        

Ingrese la cantidad de numeros pseudoaleatorios:  100
semilla 1611767638 iteracion 0
semilla 1611767638 iteracion 1
semilla 1611767638 iteracion 2
semilla 1611767638 iteracion 3
semilla 1611767638 iteracion 4
semilla 1611767638 iteracion 5
semilla 1611767638 iteracion 6
semilla 1611767638 iteracion 7
semilla 1611767638 iteracion 8
semilla 1611767638 iteracion 9
semilla 1611767638 iteracion 10
semilla 1611767638 iteracion 11
semilla 1611767638 iteracion 12
semilla 1611767638 iteracion 13
semilla 1611767638 iteracion 14
semilla 1611767638 iteracion 15
semilla 1611767638 iteracion 16
semilla 1611767638 iteracion 17
semilla 1611767638 iteracion 18
semilla 1611767638 iteracion 19
semilla 1611767638 iteracion 20
semilla 1611767638 iteracion 21
semilla 1611767638 iteracion 22
semilla 1611767638 iteracion 23
semilla 1611767638 iteracion 24
semilla 1611767638 iteracion 25
semilla 1611767638 iteracion 26
semilla 1611767638 iteracion 27
semilla 1611767638 iteracion 28
semilla 1611767638 iteracion

In [43]:
from scipy.stats import chisquare
print(lista)
chisquare(lista, ddof=0.05)

[0.2631578947368421, 0.10526315789473684, 0.42105263157894735, 0.7894736842105263, 0.05263157894736842, 0.5263157894736842, 0.5789473684210527, 0.47368421052631576, 0.6842105263157895]


Power_divergenceResult(statistic=1.1621621621621623, pvalue=0.9968424171841995)

## Resultados: 

El valor estadístico es: 1.1621621621621623 y el valor de prueba es 0.9968424171841995 por lo que los números siguen una distribución uniforme.